### DISCUSSION TITLE
watchdog_reboot delay in flashFirmware

### DISCUSSION THREAD

**Discussion by websterelec** [rhulme/pico-flashloader#11] [General]
I found that with larger programs, the delay in watchdog_reboot (500ms) isn't long enough for the initial flash_range_erase call that erases the target memory area to complete.  I measured 730ms for a size of 0x35000.  I increased the delay to 2 seconds for my application.

For some reason that I don't understand, when I had a usb connection to the pc (used for loading .uf2 files) and the watchdog expired due to insufficient time, things behaved differently and I couldn't get the board to recover.  I noticed this during panicked troubleshooting so I'm not certain if the usb connection really had an impact.

The 500ms in the original code is correct for the sample applications, but I wanted to leave a note that might be helpful to someone in the future.

**Comment by rhulme** [comment]
Hi,

Thanks for that.  Yes, if you have a large area to erase, it can take longer than 500ms.  I hadn't encountered that in the scenarios I've used it.

The cleanest solution is probably to erase smaller blocks in a loop and reset the watchdog on each pass.  Then it doesn't make any difference how big the application is.  I'll look at making that change as soon as I have time.

The issue you saw after the watchdog expired is very strange.  I'll see if I can reproduce it.

**Comment by websterelec** [comment]
I can't see how having the USB connection should impact anything.  Possibly due to my application including pico_stdlib that includes stdio.h, but I have pico_enable_stdio_usb(${PROJECTNAME} 0) in my cmakelist file.  After increasing the watchdog timeout, I did not see the issue again.  Since the USB connection won't be accessible in the final product, this would only show up for me in development.

Thanks again for sharing your code.  It has saved me a significant amount of time.

**Reply by rhulme** [reply]
I haven't tried anything yet (haven't had time this weekend unfortunately) but thinking about it, if the watchdog timer expired whilst the application was being erased, it would reboot and then determine via the scratch registers that it was in the process of performing an update and would try again, which would trigger a watchdog reset again.

The retry counter should catch that to prevent it being caught in an endless loop and end up wearing out the flash by repeatedly erasing it.  After three attempts it should drop back to the standard USB bootloader ready to accept a UF2 file.

Did you see the standard bootloader or did it just stop responding completely (i.e. maybe it *was* stuck in an endless 'erase/reboot' loop)?

**Reply by websterelec** [reply]
I have two boards I'm testing with that I believe are identical.  Board 1 does not fail with the wdt set to 500ms.  Board 2 does fail every time.  I don't know why board 1 had different results.

When board 2 fails, it falls back to the standard bootloader.  If I leave the pc connected during the time that I update the firmware, the standard bootloader window appears quickly after my applications reset.  If I disconnect the pc and repeat the process, the board does nothing until I plug the usb cable in and then the standard bootloader window appears.  In either of these cases if I cycle power to the board, my code is properly transferred.  That is, if the wdt fails, it goes to the standard bootloader, then I cycle power and my code is properly copied to the application area and the board starts normally with the new code version.

I think what I previously perceived as the board becoming non-responsive was the board sitting in the standard bootloader waiting.  I've tried repeatedly, but a power cycle always completes and loads my new version after the wd failure.  I don't understand why the wd timer doesn't fail again after the power cycle?

**Reply by rhulme** [reply]
I wondered whether maybe the erase time is shorter if the flash has already been erased.  If that were the case, as the flashloader will retry three times, more of the flash would be erased each time, but a quick test with a slightly modified 'flash_program' example on a standard Pico seemed to show a pretty consistent erase time, regardless of whether it was only erased or erased and then programmed.

Are you using the standard Pico board, or something else/custom?  Maybe other flash chips behave differently?

What size is your application (i.e. how much flash is being erased)?  It's possible that board 1 is finishing the erasing in just under 500ms and board 2 would need just a little bit more.

I'd be interested to know if increasing the number of retries appears to fix the problem.  For each power cycle, you'll get three attempts so if you increase it to six, you might find it doesn't end up in the standard bootloader.  If it really is borderline with how long it actually needs, it might be that six chances are enough for it to do it quickly enough once.

**Reply by websterelec** [reply]
We have 4 total boards that are custom and use the W25Q16JVSNIQ.  I have 2 boards and one requires the wdt timeout to be increased to operate correctly.  Someone else has the other 2 boards and is helping me test.  They found the same issue on one of their 2 boards as well.

I just tried setting the timeout back to 500 on the board I have that has the issue and I also tried increasing sMaxRetries from 3 to both 6 and 12.  It still ended up in the standard bootloader.  When I changed the timeout back to 2000 and sMaxRetries back to 3, it worked again.

I'm not sure what the root cause of the issue is, but increasing the wd timeout seems to fix it every time.  Maybe I should try to toggle a port pin during the erase and measure the erase time between the boards to see if there is actually a significant difference between them and check if the erase time is near 500.

**Reply by rhulme** [reply]
OK.  I don't know if there are any flash chips that finish more quickly if the memory was already erased but if you're using the same chip as used on the normal Pico boards, I would expect the behaviour to be the same.

I'm not sure at the moment why it should work successfully after power-cycling.  The only difference is that the flashloader would check the application's CRC, search for the update image and only then start updating rather than immediately trying to update.  I don't see anything there that would make flashing (or rather erasing) happen any faster.

It would certainly be interesting to see just how long the boards need for erasing.  Is it possible that the slower boards have been through more erase cycles?  I'm not sure myself whether one should expect erase times to increase with wear.  It's not something I've needed to investigate in the past.

The proper fix is certainly to erase a single sector, reset the watchdog and move on to the next sector.  That allows the watchdog to stay low and should work regardless of how large an area needs to be erased.

I just don't understand at the moment why power-cycling helps the boards that get reset.

**Reply by websterelec** [reply]
Thanks - I'll check the erase time asap.  Hopefully this isn't some error I've introduced.  The flashloader files haven't been changed except the wdt timeout.  My application just dumps the bin file to a location in flash, creates the header, and then resets.  If that were corrupted somehow, power-reset wouldn't fix anything.  I'll also try starting with cleared flash on both to see if that has any effect.

**Reply by websterelec** [reply]
I compared the erase times between boards.  The board that successfully programs with the timeout at 500ms takes 1.1 seconds to erase.  The board that fails and goes to the standard bootloader takes 1.2 seconds.  I changed the timeout from 500ms to 700ms and repeated with both boards and they both succeeded.  So I'm guessing I must be on the edge with the retries somehow.  For now, I'm going to leave the timeout at 2 seconds.  I haven't seen an issue with it and trying to debug further to nail down exactly what is happening hopefully isn't worth the time.

**Reply by rhulme** [reply]
OK.  I fixed the main timeout issue by erasing the flash area one block at a time and then retriggering the watchdog timer.  That should be scalable too so even if your application becomes even bigger over time, there shouldn't be any problems.

It took some fiddling but I also found out why it worked after power-cycling the board.  The hardware watchdog module needs to know how fast the main clock is running so that it can scale appropriately.  Because the flashloader wasn't initialising it, it was running at a much slower speed and therefore didn't trigger a reset (the time take to erase flash is independent of the speed the RP2040 is running at).

If the normal application is started and this uses the standard SDK initialisation code, the watchdog is initialised correctly so the timeout was correct for every warm start afterwards.

On a cold start where there was no valid application, the timeout would be much longer and the update process had time to run through.

**Reply by websterelec** [reply]
That makes sense.  Thanks for digging into it.  I'll update my code with the new versions and let you know how it goes after I've done some testing.  I'm using your code to allow both firmware updates locally, through a USB interface, and remotely, via cloud transfer.  Remote updates are both comforting (fix bugs remotely) and concerning (brick every unit in the field).

**Reply by rhulme** [reply]
When you say "bricking every unit in the field", do you mean there is the potential to brick them, or that they are being bricked?
There's not a huge amount that a flashloader can do to stop you flashing buggy software.  It can be theoretically mitigated by having some kind of fallback mechanism that either flashes the previous "known good" software or if the flashloader itself can have the ability to fetch a new update from somewhere but the main problem is determining that the new software has a problem.  There are any number of scenarios where problems _could_ be caught and just as many where it it's incredibly difficult to somehow programmatically recognise issues.  Then it mainly comes down to testing as much as possible before deploying and managing the roll-out so not every device immediately updates to the latest software.

If you don't already, I would seriously consider using the 'urloader' branch to allow you to also update the flashloader in the field if required.

**Reply by websterelec** [reply]
When I said "bricking every unit in the field", I'm just worrying that something could happen during the update process that causes a failure where the board locks up or ends up in the standard bootloader.  I understand that should never happen based on how your code works.  It has a a checksum to make sure the new code is valid and if power is lost at any point during the update, it will just start over when power is restored.  I've updated firmware hundreds of times locally through a USB drive and the only issue I had was the wd timeout.  In case of bugs in my application, I am testing as much as possible and intend to only allow updates on a site-by-site basis initially to make sure things go smoothly.  I'll check out the urloader branch and hope to have time to implement it.  Thanks for all your help.
